Sidebar info view

iainmcampbell opened the pull request:
- https://github.com/bikespace/bikespace/issues/209

- Implemented redesigned sidebar tabs
- Added info panel

<table>
<tr><td colspan="3">
<img width="1175" alt="Screenshot 2025-04-16 at 10 31 59 PM" src="https://github.com/user-attachments/assets/248e9b9d-8ffd-4168-b267-866692740e85" />
</td></tr><tr><td colspan="3">
<img width="1069" alt="Screenshot 2025-04-16 at 9 56 55 PM" src="https://github.com/user-attachments/assets/5424edcb-762f-48ee-a80b-98dad1eb8e40" />
</td></tr><tr><td>
<img width="377" alt="Screenshot 2025-04-16 at 9 57 06 PM" src="https://github.com/user-attachments/assets/21290ea5-808c-42bf-9102-d9182762244f" />
</td><td>
<img width="373" alt="Screenshot 2025-04-16 at 10 32 18 PM" src="https://github.com/user-attachments/assets/bb533d75-d4f7-4ddc-bf2c-db4ac851537a" />
</td><td>
<img width="374" alt="Screenshot 2025-04-16 at 9 57 36 PM" src="https://github.com/user-attachments/assets/d26dbc64-471a-4d5d-ac44-511a60dd5747" />
</td></tr></table>

I’ve tried to refactor the sidebar styles to be mobile-first.

I used SVGR (svg as jsx file) for the sidebar icons so we can style them using CSS.

Since the server doesn't know whether we're mobile or not, this new pattern where the sidebar content is open by default on desktop and closed by default on mobile is a bit of a challenge to implement using server-side rendering. I made it work by defaulting the sidebar content to closed, so we never render it on the server. Once the client is hydrated, we can use a media query to know we're in a desktop viewport, and then render the sidebar content. Let me know if this makes sense, or if there’s a better way to handle this.

I tried to use the colour variables in [variables.module.scss](https://github.com/bikespace/bikespace/blob/main/bikespace_frontend/src/styles/variables.module.scss), they might not match the Figma 100% but we should be sticking to a standard colour palette, right?

I tried to use the existing SidebarButton component, which is close to but doesn't quite match the mockups.

## Questions

Where is the data guide PDF?

I assume the 'Data last updated' in the info view is meant to represent the date of the most recent submission? I tried to piggyback off the Feeds view, but that data isn't always available. I can make a new query - does the backend have a way to fetch only the most recent submission? This date is also available in the Feed view, so I don't know if it’s worth setting up a new endpoint just for this.

iainmcampbell committed:
Move sidebar tabs to left side, add icons

iainmcampbell committed:
Mobile sidebar + be able to close sidebar content

tallcoleman commented:
Thanks for taking this on @iainmcampbell ! 

Just wanted to check since I don't understand this note:

> having the sidebar content open by default on desktop and closed by default on mobile is a bit of a challenge, the server doesn't know whether we're mobile or not

Which server needs to know the client state? The way we deploy is to build the site to static so there's no server used for frontend content, only the 'backend' server to get the data from the API.

iainmcampbell committed:
Rename sidebar data -> insights

iainmcampbell committed:
Consolidate sidebar content, add info panel

iainmcampbell commented:
> Which server needs to know the client state? The way we deploy is to build the site to static so there's no server used for frontend content, only the 'backend' server to get the data from the API.

Ah, gotcha. I was getting SSR errors about media queries, I assume because next.js does SSR in dev mode. but a static build would have the same issue - it _very_ definitely cannot know anything about the client’s computer, eg whether they're on a phone or not.

These mockups have client state working differently in mobile vs desktop. Previously the mobile view just had different CSS, but now in mobile we have different state too. Now we don't render any sidebar tab content by default, and introduce a new concept of "opening" a sidebar tab, which we don't have on desktop. I got around this by not opening a sidebar tab by default, and then once the page is rendered, check if we're on desktop, and then open the panel.

I wanted to mention it in case there’s a more elegant way to solve this problem.

iainmcampbell commented:
ok I just tried the classic "is it SSR" test of loading bikespace.ca with JS disabled, and like you said the static build doesn't do server-side rendering in production. it's just `<body><script></body>`.

So it's puzzling that it's trying to do it in development. if we can turn off SSR in dev, we could remove a bunch of future complexity.

I don't have a lot of experience with next.js so I may be misunderstanding here?

sonalranjit commented:
@iainmcampbell to build the site statically on your local environment i believe you can run ```npm build``` or ```npm start``` i believe but im not a 100% confident.
If you run it in develop mode with hot reloading i believe it runs with SSR.
Also im not sure if you saw the preview link, sorry it's not too obvious in the checks, your PR change are deployed here https://48cd184a.bikespace-v2.pages.dev/ 
Link is hidden underneath the Cloudflare Pages block.

tallcoleman commented on bikespace_frontend/src/assets/images/data-guide.png:
Do we have the svg versions of these saved somewhere?

iainmcampbell replied:
I can use SVGs for these instead if that would be better, I just exported them from Figma

tallcoleman replied:
OK thanks!

tallcoleman commented on bikespace_frontend/src/components/dashboard/sidebar-tab-content/_SidebarContentInfo.tsx:
```diff
@@ -0,0 +1,44 @@
+  lastUpdated: 'Data Last Updated',
+  about: 'About the Dashboard',
+  aboutContent:
+    'The BikeSpace Dashboard is based on issues reported through the publicly accessible Bikespace Issue Reporting web app.\nLaunched in July 2018, BikeSpace allows cyclists to self-report bike parking concerns they come across in the Toronto, Canada area.',
```
What's the benefit of putting the text content here instead of just writing it into the return function below?

Also, the `\n` in `aboutContent` does not get rendered FYI. You could either make it into a list and use `map` to render each value as its own paragraph or put the content into an MDX file that gets imported instead. (When I tested this, in `tsconfig.json`, I had to add `"mdx"` to `types` and `"**/*.mdx"` to `include`).

iainmcampbell replied:
old habits. if you're translating your site, you generally have lint rules to prevent strings in the JSX. if that's not a concern here, then I’ll just put the strings in the JSX, then no need to jump through hoops to get newlines to render.

tallcoleman replied:
Right - translation! makes total sense. I think in this case, strings in the JSX is fine.

tallcoleman commented on bikespace_frontend/src/components/dashboard/sidebar-tabs/icons/export.tsx:
An a11y challenge with using these svgs on their own (instead of as the src for an img tag) is that you can't set `alt=""` to indicate that the image is decorative. This got flagged when I tested with the Firefox accessibility tools. The equivalent options for svg don't seem to have a way to indicate the icon is decorative either.

One option would be to add `role="img"` in the top-level `svg` tag and a title tag (e.g. `<title>Export</title>` before the `path` tag). In theory this might mean a screen reader reads the nav title twice (once via the svg title tag, and once via the label on the button element). Though when I tested it with VoiceOver, it just read the button label.

tallcoleman replied:
Update: I looked into this some more:

- [The solution](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Roles/img_role#svg_and_roleimg) for a non-presentational svg is to add `role="img"` and an `aria-label` to the svg. In this case though, I think the right solution would be to add `role="presentation"`, which is equivalent to an img tag with `alt=""`.
- The reason that VoiceOver isn't reading the children of `button` is that the [ARIA button role treats all descendents as presentational](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Roles/button_role#all_descendants_are_presentational). So the Firefox accessibility tools are probably flagging it incorrectly, but adding `role="presentation"` should fix it just in case other screen readers or browsers handle `button` slightly differently.

tallcoleman committed:
Create Bikespace Dashboard Dataset Details.pdf

tallcoleman reviewed (changes requested):
Thanks @iainmcampbell this is really great! Very exciting to have this implemented.

I'll note the general comments I jotted down during testing; if you want me to help with any of these, just let me know and I'll push some commits to the branch.

- The Info, Insights, and Filters tabs did not scroll for me on desktop when the content exceeded the viewport height.
- The line spacing for the text on the Info tab (170%?) seems quite generous compared to other body text (e.g. comment text on the Feed tab; text on the About page, which is 140%)
- BikeSpace should be property capitalized everywhere it is used
- I pushed a commit to add the data dictionary PDF to the /public folder - please add link to this PDF in the button on the Info tab
- The Feed tab serves to show pop-up text on mobile, so it should actually be half-height on mobile when opened (see screenshot from the Figma design below)
- The width of the dashboard nav is _slightly_ different when Feed/Info are selected s. when Insights/Filters are selected.

There is also a bug where the dashboard navigation stops working if a map pin is selected. This appears to be an existing bug, so if you don't want to address that in this PR, I'll open an issue and fix it after this is merged.

Thanks again and let me know if you have any questions!

<img width="321" alt="Screenshot 2025-04-18 at 8 49 53 AM" src="https://github.com/user-attachments/assets/90be1bde-614a-484c-ac27-86264399b822" />

tallcoleman commented:
Sorry, one more thing I forgot - similar to the mobile styling when a tab is open, there needs to be padding below the dashboard nav on mobile so that the buttons don't conflict with the OS UI elements. e.g. see screenshot below on an iPhone running the site as a PWA.

<img src="https://github.com/user-attachments/assets/8491fc1b-faf7-4291-b142-624f442fc4ed" width=250 />

iainmcampbell committed:
Avoid SSR in development -> simplify useIsMobile hook

iainmcampbell committed:
fix scroll, clean up naming, hopefully less confusing

iainmcampbell committed:
Feed tab fixes

iainmcampbell commented:
thanks for a very thorough review!!

not done yet, but I believe I've addressed the following:

> - The Info, Insights, and Filters tabs did not scroll for me on desktop when the content exceeded the viewport height.
> - The line spacing for the text on the Info tab (170%?) seems quite generous compared to other body text (e.g. comment text on the Feed tab; text on the About page, which is 140%)
> - BikeSpace should be property capitalized everywhere it is used
> - I pushed a commit to add the data dictionary PDF to the /public folder - please add link to this PDF in the button on the Info tab
> - The Feed tab serves to show pop-up text on mobile, so it should actually be half-height on mobile when opened (see screenshot from the Figma design below)
> - The width of the dashboard nav is slightly different when Feed/Info are selected s. when Insights/Filters are selected.

tallcoleman commented:
> not done yet, but I believe I've addressed the following:

Thanks @iainmcampbell , appreciate it! Let us know when you want someone to take a final look.

iainmcampbell committed:
final tweaks

iainmcampbell commented:
Ok, I think I got everything:

- added padding below mobile dashboard nav
- added `role="presentation"` to the SVG icons `<svg>` elements
- converted the info page images to SVGs

should also note I loaded the Sidebar component dynamically using Next.js’s `dynamic()` function, which seems to be the least hack-y way to avoid SSR. This allowed me to remove a bunch of logic (https://github.com/bikespace/bikespace/pull/249/commits/56190eab28b82113bb5cd0be991f387504b077b0)

tallcoleman committed:
Capitalization fixes

tallcoleman committed:
Prevent sidebar from shrinking to content size on desktop

tallcoleman committed:
Rename Bikespace Dashboard Dataset Details.pdf

tallcoleman committed:
Fix sidebar content height on Safari

Safari vh includes browser UI not just area the user can actually view - have to add backup rule using dvh which only measures viewable area. dvh still not >95% supported, hence why you still have to keep the vh rule as well.

tallcoleman committed:
Merge branch 'sidebar-info-view' of https://github.com/bikespace/bikespace into sidebar-info-view

tallcoleman commented:
@iainmcampbell only one bug to fix and then this is ready to go:

For some reason on Safari, when the sidebar content exceeds the available height and has to scroll, the button disappears (examples with different heights below):

<img width="394" alt="Screenshot 2025-05-03 at 9 21 10 AM" src="https://github.com/user-attachments/assets/100a1548-7247-4d9b-92de-a7be7f777000" />

<img width="294" alt="Screenshot 2025-05-03 at 9 22 30 AM" src="https://github.com/user-attachments/assets/7de19e85-0eff-400e-8e36-c86133104075" />

I've been unable to puzzle this one out, so if you have any ideas, let me know!

I was able to fix another bug I've run into before with using `vh` on Safari and made a couple other very small style tweaks.

tallcoleman committed:
Bugfix - mobile layout and scrolling

- fixed issue with return to map button not showing on Safari

tallcoleman committed:
Updated style for feed content heading

Increase usable space by reducing padding

tallcoleman reviewed (approved):
@iainmcampbell thanks again for all your help on this! I think I fixed the bug with the back to map button not showing on Safari. Take a look and if it all seems OK with you, feel free to merge.

iainmcampbell commented:
@tallcoleman nice catch, fix looks great!
